ci: division by zero test failing for pyspark

the test introduced in https://github.com/narwhals-dev/narwhals/pull/2636 fails for pyspark, which isn't run by default

we should update the pyspark code to use `try_divide`

---

https://github.com/narwhals-dev/narwhals/actions/runs/15945539506/job/44979105403

```
_________________ test_floordiv_int_by_zero[pyspark-0-0-None] __________________

left = 0, right = 0, expected = None
constructor = <function pyspark_lazy_constructor.<locals>._constructor at 0x7fe8efc98180>
request = <FixtureRequest for <Function test_floordiv_int_by_zero[pyspark-0-0-None]>>

    @pytest.mark.parametrize(
        ("left", "right", "expected"),
        [(-2, 0, float("-inf")), (0, 0, None), (2, 0, float("inf"))],
    )
    @pytest.mark.skipif(PANDAS_VERSION < (2, 0), reason="converts floordiv by zero to 0")
    def test_floordiv_int_by_zero(
        left: int,
        right: int,
        expected: float | None,
        constructor: Constructor,
        request: pytest.FixtureRequest,
    ) -> None:
        data: dict[str, list[int]] = {"a": [left]}
        df = nw.from_native(constructor(data))
        # pyarrow backend floordiv raises divide by zero error
        # ibis backend floordiv cannot cast value to inf or -inf
        if any(x in str(constructor) for x in ["ibis", "pyarrow"]):
            request.applymarker(pytest.mark.xfail)
        # duckdb backend floordiv return None
        if "duckdb" in str(constructor):
            floordiv_result = df.select(nw.col("a") // right)
            assert_equal_data(floordiv_result, {"a": [None]})
        # polars backend floordiv returns null
        elif "polars" in str(constructor) and "lazy" not in str(constructor):
            floordiv_result = df.select(nw.col("a") // right)
            assert all(floordiv_result["a"].is_null())
        # polars lazy floordiv cannot be sliced and returns None
        elif all(x in str(constructor) for x in ["polars", "lazy"]):
            floordiv_result = df.select(nw.col("a") // right)
            assert_equal_data(floordiv_result, {"a": [None]})
        # pandas[nullable] backend floordiv always returns 0
        elif all(x in str(constructor) for x in ["pandas", "nullable"]):
            floordiv_result = df.select(nw.col("a") // right)
            assert_equal_data(floordiv_result, {"a": [0]})
        else:
            floordiv_result = df.select(nw.col("a") // right)
>           assert_equal_data(floordiv_result, {"a": [expected]})

tests/expr_and_series/division_by_zero_test.py:115: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/utils.py:92: in assert_equal_data
    result = result.collect(**kwargs.get(result.implementation, {}))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
narwhals/dataframe.py:2336: in collect
    self._compliant_frame.collect(backend=eager_backend, **kwargs), level="full"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
narwhals/_spark_like/dataframe.py:234: in collect
    self._collect_to_arrow(),
    ^^^^^^^^^^^^^^^^^^^^^^^^
narwhals/_spark_like/dataframe.py:196: in _collect_to_arrow
    return self.native.toArrow()
           ^^^^^^^^^^^^^^^^^^^^^
/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/pyspark/sql/classic/dataframe.py:1789: in toArrow
    return PandasConversionMixin.toArrow(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/pyspark/sql/pandas/conversion.py:249: in toArrow
    batches = self._collect_as_arrow(
/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/site-packages/pyspark/sql/pandas/conversion.py:315: in _collect_as_arrow
    with unwrap_spark_exception():
/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/contextlib.py:158: in __exit__
    self.gen.throw(typ, value, traceback)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    @contextmanager
    def unwrap_spark_exception() -> Iterator[Any]:
        from pyspark import SparkContext
        from py4j.protocol import Py4JJavaError
        from py4j.java_gateway import is_instance_of
    
        assert SparkContext._gateway is not None
    
        gw = SparkContext._gateway
        try:
            yield
        except Py4JJavaError as e:
            je: "Py4JJavaError" = e.java_exception
            if je is not None and is_instance_of(gw, je, "org.apache.spark.SparkException"):
                converted = convert_exception(je.getCause())
                if not isinstance(converted, UnknownException):
>                   raise converted from None
E                   pyspark.errors.exceptions.captured.ArithmeticException: [DIVIDE_BY_ZERO] Division by zero. Use `try_divide` to tolerate divisor being 0 and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error. SQLSTATE: 22012
E                   == DataFrame ==
E                   "__truediv__" was called from
E                   /home/runner/work/narwhals/narwhals/narwhals/_spark_like/expr.py:403
```

--- a/narwhals/_spark_like/expr.py
+++ b/narwhals/_spark_like/expr.py
@@ -19,6 +19,7 @@ from narwhals._spark_like.utils import (
     import_native_dtypes,
     import_window,
     narwhals_to_native_dtype,
+    true_divide,
 )
 from narwhals._utils import Implementation, not_implemented, parse_version
 from narwhals.dependencies import get_pyspark
@@ -398,15 +399,27 @@ class SparkLikeExpr(LazyExpr["SparkLikeLazyFrame", "Column"]):
             implementation=self._implementation,
         )
 
+    def __truediv__(self, other: SparkLikeExpr) -> Self:
+        def _truediv(expr: Column, other: Column) -> Column:
+            return true_divide(self._F, expr, other)
+
+        return self._with_binary(_truediv, other)
+
+    def __rtruediv__(self, other: SparkLikeExpr) -> Self:
+        def _rtruediv(expr: Column, other: Column) -> Column:
+            return true_divide(self._F, other, expr)
+
+        return self._with_binary(_rtruediv, other).alias("literal")
+
     def __floordiv__(self, other: SparkLikeExpr) -> Self:
         def _floordiv(expr: Column, other: Column) -> Column:
-            return self._F.floor(expr / other)
+            return self._F.floor(true_divide(self._F, expr, other))
 
         return self._with_binary(_floordiv, other)
 
     def __rfloordiv__(self, other: SparkLikeExpr) -> Self:
         def _rfloordiv(expr: Column, other: Column) -> Column:
-            return self._F.floor(other / expr)
+            return self._F.floor(true_divide(self._F, other, expr))
 
         return self._with_binary(_rfloordiv, other).alias("literal")
 

--- a/narwhals/_spark_like/namespace.py
+++ b/narwhals/_spark_like/namespace.py
@@ -16,6 +16,7 @@ from narwhals._spark_like.utils import (
     import_functions,
     import_native_dtypes,
     narwhals_to_native_dtype,
+    true_divide,
 )
 
 if TYPE_CHECKING:
@@ -152,20 +153,14 @@ class SparkLikeNamespace(
         def func(cols: Iterable[Column]) -> Column:
             cols = list(cols)
             F = exprs[0]._F  # noqa: N806
-            # PySpark before 3.5 doesn't have `try_divide`, SQLFrame doesn't have it.
-            divide = getattr(F, "try_divide", operator.truediv)
-            return divide(
-                reduce(
-                    operator.add, (self._F.coalesce(col, self._F.lit(0)) for col in cols)
-                ),
-                reduce(
-                    operator.add,
-                    (
-                        col.isNotNull().cast(self._native_dtypes.IntegerType())
-                        for col in cols
-                    ),
-                ),
+            numerator = reduce(
+                operator.add, (self._F.coalesce(col, self._F.lit(0)) for col in cols)
             )
+            denominator = reduce(
+                operator.add,
+                (col.isNotNull().cast(self._native_dtypes.IntegerType()) for col in cols),
+            )
+            return true_divide(F, numerator, denominator)
 
         return self._expr._from_elementwise_horizontal_op(func, *exprs)
 

--- a/narwhals/_spark_like/utils.py
+++ b/narwhals/_spark_like/utils.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import operator
 from functools import lru_cache
 from importlib import import_module
 from typing import TYPE_CHECKING, Any, overload
@@ -283,3 +284,9 @@ def strptime_to_pyspark_format(format: str | None) -> str | None:
     for py_format, spark_format in DATETIME_PATTERNS_MAPPING.items():
         pyspark_format = pyspark_format.replace(py_format, spark_format)
     return pyspark_format.replace("T", " ")
+
+
+def true_divide(F: Any, left: Column, right: Column) -> Column:  # noqa: N803
+    # PySpark before 3.5 doesn't have `try_divide`, SQLFrame doesn't have it.
+    divide = getattr(F, "try_divide", operator.truediv)
+    return divide(left, right)

--- a/tests/expr_and_series/division_by_zero_test.py
+++ b/tests/expr_and_series/division_by_zero_test.py
@@ -70,7 +70,7 @@ def test_series_floordiv_int_by_zero(
 def test_truediv_by_zero(
     left: float, right: float, expected: float | None, constructor: Constructor
 ) -> None:
-    if "pyspark" in str(constructor) and "sqlframe" not in str(constructor):
+    if "pyspark" in str(constructor):
         # https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.try_divide.html
         # PySpark always returns null when dividing by zero.
         expected = None
@@ -99,7 +99,7 @@ def test_floordiv_int_by_zero(
     if any(x in str(constructor) for x in ["ibis", "pyarrow"]):
         request.applymarker(pytest.mark.xfail)
     # duckdb backend floordiv return None
-    if "duckdb" in str(constructor):
+    if any(x in str(constructor) for x in ("duckdb", "pyspark")):
         floordiv_result = df.select(nw.col("a") // right)
         assert_equal_data(floordiv_result, {"a": [None]})
     # polars backend floordiv returns null
@@ -111,7 +111,7 @@ def test_floordiv_int_by_zero(
         floordiv_result = df.select(nw.col("a") // right)
         assert_equal_data(floordiv_result, {"a": [None]})
     # pandas[nullable] backend floordiv always returns 0
-    elif all(x in str(constructor) for x in ["pandas", "nullable"]):
+    elif "pandas_nullable" in str(constructor):
         floordiv_result = df.select(nw.col("a") // right)
         assert_equal_data(floordiv_result, {"a": [0]})
     else:

--- a/tests/expr_and_series/division_by_zero_test.py
+++ b/tests/expr_and_series/division_by_zero_test.py
@@ -70,6 +70,10 @@ def test_series_floordiv_int_by_zero(
 def test_truediv_by_zero(
     left: float, right: float, expected: float | None, constructor: Constructor
 ) -> None:
+    if "pyspark" in str(constructor) and "sqlframe" not in str(constructor):
+        # https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.try_divide.html
+        # PySpark always returns null when dividing by zero.
+        expected = None
     data: dict[str, list[int | float]] = {"a": [left]}
     df = nw.from_native(constructor(data))
     truediv_result = df.select(nw.col("a") / right)

